### DISCUSSION TITLE
Control TRV via Cloud

### DISCUSSION THREAD

**Discussion by hollenal** [windkh/node-red-contrib-shelly#192] [Q&A] [answered]
Has anyone managed to control a TRV or BLU TRV via the Cloud node? I cannot find any documentation on the correct command structure.

**Comment by windkh** [comment, marked as answer]
@hollenal you could not find the command structure because there is none — the cloud node has no thermostat support at all, so no payload can work. Sorry that this sat unanswered for so long.

The node dispatches on `msg.payload.type` and handles exactly six values:

| `type` | Cloud endpoint |
| --- | --- |
| `light` | `/device/light/control` |
| `relay` | `/device/relay/control` |
| `roller` | `/device/relay/roller/control` |
| `relays` | `/device/relay/bulk_control` |
| `status` | `/device/status` |
| `all_status` | `/device/all_status` |

Anything else falls through and the message is passed on untouched, which is why it looks like a silent no-op rather than an error.

## What does work

**Reading a TRV over the cloud** is fine — the status endpoint returns whatever the device reports, thermostat fields included:

```json
{ "type": "status", "id": "<your-device-id>" }
```

**Controlling a Shelly TRV** works locally today. Use a `shelly-gen1` node with the device type set to Shelly TRV (`SHTRV-01`, typed `Thermostat`) and send:

```json
{ "position": 50 }
{ "temperature": 21.5 }
{ "schedule": true }
{ "scheduleProfile": 2 }
{ "boostMinutes": 30 }
```

`position` accepts 0–100, `temperature` 4–31 °C, `scheduleProfile` 1–5; values outside those ranges are dropped rather than sent. Several can be combined in one message.

**A BLU TRV is a different matter.** It has no IP address at all — it is a Bluetooth device reached through a gen2+ device acting as a BLU gateway, so the cloud node was never the route for it. This package surfaces BLU *events* on the gateway node (enable callback mode plus the bluetooth checkbox on the server config node); BLU TRV *control* would go through the gateway rather than the cloud.

## Tracking the gap

Thermostat support in the cloud node is a genuine hole — relays, lights and rollers are covered and thermostats are not. I have opened #282 for it so it does not get lost again in a discussion thread.

**Comment by windkh** [comment]
Following up on my own answer — I said a BLU TRV goes through the gateway rather than the cloud, but left it there. Here is the actual payload, @hollenal, and it needs no new code.

## Controlling a BLU TRV today

A BLU TRV is driven through its gateway with [`BluTrv.call`](https://shelly-api-docs.shelly.cloud/docs-ble/Devices/BLU_ZB/trv/). The `BluGateway` device type routes to the generic gen2 parser, so any RPC method passes straight through. Send this to a `shelly-gen2` node bound to your BLU gateway:

```json
{
  "method": "BluTrv.call",
  "params": {
    "id": 200,
    "method": "TRV.SetTarget",
    "params": { "id": 0, "target_C": 21.5 }
  }
}
```

- The **outer** `id` (`200` above) is the TRV's component id on the gateway — check yours, it is not always 200. `BluTrv.GetRemoteStatus` or the gateway's web UI will tell you.
- The **inner** `id` is the TRV component on the device itself, normally `0`.
- `target_C` is the target temperature in °C.

`Trv.SetConfig` works the same way for `default_override_target_C` and the `min_target_C` / `max_target_C` bounds.

I ran that message through the node's parser to confirm rather than assume; it produces exactly the JSON-RPC frame the gateway expects:

```json
{
  "route": "/rpc",
  "method": "POST",
  "data": { "id": 1, "method": "BluTrv.call",
            "params": { "id": 200, "method": "TRV.SetTarget", "params": { "id": 0, "target_C": 21.5 } } }
}
```

## And a correction on the cloud side

My earlier answer said the cloud node has no thermostat support. That is right, but I have since checked the API itself, and the reason matters: **the Shelly Cloud Control API has no thermostat endpoint at all** — not in the [deprecated v1](https://shelly-api-docs.shelly.cloud/cloud-control-api/communication) (relay, roller, light, bulk, status) and not in the [v2.0-beta](https://shelly-api-docs.shelly.cloud/cloud-control-api/communication-v2) replacement (`/set/switch`, `/set/cover`, `/set/light`, `/set/groups`). Neither offers a generic RPC passthrough either.

So this is not something I can add — there is no endpoint to call. Cloud TRV control would have to come from Shelly first. Tracked in #282, where I have also noted that the node ought to *tell* you when it does not recognise a command instead of silently forwarding the message, which is what made your question hard to answer in the first place.

For a wired Shelly TRV (`SHTRV-01`) the gen1 node covers it locally, as in my previous reply.
